Lecture3 c

SehedaOleh opened the pull request:
In general I try to accumulate all previous notes and create this task more precise. Could you please review it and send me some feedback. thanks a lot.

SehedaOleh committed:
Lecure 3 ex1 done.

SehedaOleh committed:
Function to find the biggest element in integer array is done.

SehedaOleh committed:
Task 3 done. Add c-file and header file with functions.

rikorsev commented on Lecture_3/L3_ex1.c:
```diff
@@ -0,0 +1,44 @@
+int main()
+{
+    char initial_string[256];
+    char digital_string[256] = { };
```
It is good that you have initialized the array, but I would prefer that you do it in more explicit way:
char digital_string[256] = { 0 };

rikorsev commented on Lecture_3/L3_ex1.c:
```diff
@@ -0,0 +1,44 @@
+    char digital_string[256] = { };
+    uint8_t str_length = 0;
+    
+    scanf( "%s" , &initial_string );
```
Here is two mistakes:
1) scanf is not the best function to get a string, because if you write string like:
"my numbers are 11 and 56"
**initial_string** would contains only "my" string.
in this case is much more better to use **gets** function

2) **initial_string** - is already a pointer, so you haven't to pass to the function address of a pointer but only address that this pointer stores.
in your case the code is works but if you allocate **initial_string** like this:
char initial_string = malloc(256);
then address of pointer and address that the pointer are contains would be different and it could lead to unpredictable consequences.
The proper way  here is:
scanf( "%s" , initial_string );

rikorsev commented on Lecture_3/L3_ex2.c:
```diff
@@ -0,0 +1,48 @@
+#include <stdint.h>
+#include <time.h>
+
+#define array_length 25
```
Usual C code convention says that all macroses should be written with upper case like this:
#define ARRAY_LENGTH 25

rikorsev commented on Lecture_3/L3_ex2.c:
```diff
@@ -0,0 +1,48 @@
+#include <string.h>
+#include <stdint.h>
+#include <time.h>
+
```
**rand** function is declared in **stdlib.h** header. You forgot to add it here.

rikorsev commented on Lecture_3/L3_ex2.c:
```diff
@@ -0,0 +1,48 @@
+        printf("%d ", digit_array[i]);
+    } 
+    
+    max_number = digit_array[0];
```
here is unnecessary assignment

rikorsev commented on Lecture_3/ListLib.h:
```diff
@@ -0,0 +1,124 @@
+/***************************************
```
Never put function implementation in header files (.h)! Put here only reference to function.
If you try to include **ListLib.h** in one more file it will lead to errors on linking stage, because linker would treat that program has two same functions **ListAdd** for example. Also, it is good to protect you code from double including like this:
#pragma once
or 
#ifndef LIST_LIB_H
#define LIST_LIB_H
/* ...header content... */
#endif

rikorsev commented on Lecture_3/L3_ex3.c:
```diff
@@ -0,0 +1,63 @@
+    
+    list_init = malloc(sizeof(link_list_t));
+    
+    if (list_init == NULL)
```
Good! You check result of malloc function!

rikorsev commented on Lecture_3/ListLib.h:
```diff
@@ -0,0 +1,124 @@
+{
+    int number;
+    float digit;
+    struct link_list *new_element;
```
it is better to call this member as **next** or **next element**

rikorsev commented on Lecture_3/ListLib.h:
```diff
@@ -0,0 +1,124 @@
+
+    // in the end of current list adding new item
+    
+    current_list -> new_element = malloc(sizeof(link_list_t));
```
And here you don't check result of malloc :(

rikorsev commented on Lecture_3/ListLib.h:
```diff
@@ -0,0 +1,124 @@
+function ListShows: shows all items in the list
+
+***************************************/
+void ListShows (link_list_t *ptr_list)
```
Why do you split showing in two different loops?

SehedaOleh replied:
It was done only for visual reason. I can use only one loop for this.

rikorsev commented on Lecture_3/ListLib.h:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    // in the end of current list removing item
+    value_d = current_list -> digit;
```
what is the purpose of this variables and this assignment?

SehedaOleh committed:
Added changes according Alexandr comments.
Added lib stdlib.h.
Change define to upper case.
Removed : "max_number = digit_array[0];"

SehedaOleh committed:
Ex1 edited. Removed scanf() to gets().

SehedaOleh committed:
Added changes to Ex3 and ListLib.h

SehedaOleh committed:
Removed all functions implementation from *.h file to *.c file.

SehedaOleh commented:
Hello Alexander. I updated task according to yours comment.

SehedaOleh committed:
Changed file from *.c.txt to *.c

rikorsev commented on Lecture_3/L3_ex1.c:
```diff
@@ -0,0 +1,45 @@
+            str_index++;
+        }
+    }
+}
```
Good.

rikorsev commented on Lecture_3/L3_ex2.c:
```diff
@@ -0,0 +1,47 @@
+        
+    }
+    return max;
+}
```
Good

rikorsev commented on Lecture_3/L3_ex3.c:
```diff
@@ -0,0 +1,63 @@
+    }
+    printf("\n");
+    
+    /* * * * * * debug * * * * * * * * * * * * * * * * * *
```
If you need to hide some debug stuff it is better to use the following construction:
#ifdef DEBUG
/* some implementation */
#endif

rikorsev commented on Lecture_3/ListLib.c:
```diff
@@ -0,0 +1,118 @@
+function ListShows: shows all items in the list
+
+***************************************/
+void ListShows (link_list_t *ptr_list)
```
I still don't understand why do you need two separate **while** loops here?

rikorsev commented on Lecture_3/ListLib.c:
```diff
@@ -0,0 +1,118 @@
+void ListRemove (link_list_t *ptr_list)
+{
+    link_list_t * current_list = ptr_list;
+    //int value_n;
```
It is bad idea to leave commented code in the files

rikorsev reviewed (approved):
In general pool request looks more/less ok. Just few blots here. Clear it and you could merge the pull request.
